packages/medical-domain/src/domains/repositories/statistics/insulin-statistics.service.ts

```diff
@@ -0,0 +1,111 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import type DateFilter from '../../models/time/date-filter.model'
+import type Bolus from '../../models/medical/datum/bolus.model'
+import type Basal from '../../models/medical/datum/basal.model'
+import BasalService from '../medical/datum/basal.service'
+import BolusService from '../medical/datum/bolus.service'
+import {
+  type BasalBolusStatistics,
+  type TotalInsulinAndWeightStatistics
+} from '../../models/statistics/basal-bolus-statistics.model'
+import { getWeekDaysFilter } from './statistics.utils'
+import type PumpSettings from '../../models/medical/datum/pump-settings.model'
+import { type ParameterConfig } from '../../models/medical/datum/pump-settings.model'
+
+function resamplingDuration(basals: Basal[], start: number, end: number): Basal[] {
+  return basals.map(basal => {
+    if (basal.epoch < start) {
+      basal.epoch = start
+    }
+    if (basal.epochEnd > end) {
+      basal.epochEnd = end
+    }
+    basal.duration = basal.epochEnd - basal.epoch
+    return basal
+  })
+}
+
+function getWeight(allPumpSettings: PumpSettings[]): ParameterConfig | undefined {
+  const lastPumpSettings = allPumpSettings[allPumpSettings.length - 1]
+
+  if (!lastPumpSettings) {
+    return undefined
+  }
+
+  const weight = lastPumpSettings.payload.parameters.find(parameter => parameter.name === 'WEIGHT')
+  return weight ?? undefined
+}
+
+function getBasalBolusData(basalsData: Basal[], bolus: Bolus[], numDays: number, dateFilter: DateFilter): BasalBolusStatistics {
+  if (numDays === 0) {
+    return {
+      bolus: NaN,
+      basal: NaN,
+      total: NaN
+    }
+  }
+
+  const filteredBasal = BasalService.filterOnDate(basalsData, dateFilter.start, dateFilter.end, getWeekDaysFilter(dateFilter))
+  const filteredBolus = BolusService.filterOnDate(bolus, dateFilter.start, dateFilter.end, getWeekDaysFilter(dateFilter))
+  const basalData = resamplingDuration(filteredBasal, dateFilter.start, dateFilter.end)
+  const bolusTotal = filteredBolus.reduce((accumulator, bolus) => accumulator + bolus.normal, 0)
+  const basalTotal = basalData.reduce((accumulator, basal) => accumulator + (basal.duration / 3_600_000 * basal.rate), 0)
+  const totalBasalBolus = bolusTotal + basalTotal
+
+  return {
+    bolus: bolusTotal / numDays,
+    basal: basalTotal / numDays,
+    total: totalBasalBolus / numDays
+
+  }
+}
+
+function getTotalInsulinAndWeightData(basalsData: Basal[], bolusData: Bolus[], numDays: number, dateFilter: DateFilter, pumpSettings: PumpSettings[]): TotalInsulinAndWeightStatistics {
+  const weight = getWeight(pumpSettings)
+  const { basal, bolus } = getBasalBolusData(basalsData, bolusData, numDays, dateFilter)
+  const totalInsulin = [basal, bolus].reduce((accumulator, value) => {
+    const delivered = isNaN(value) ? 0 : value || 0
```

Comment:
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches to further communicate your intent.  -->
**🎯 suggestion (non-blocking):** ‏I don't understand the logic, if the value is NaN or undefined we want to return 0 so why not do:

```ts
const delivered = isNaN(value) || !value ? 0 : value
```
(maybe I'm missing something here)